core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -92,7 +92,7 @@ private enum State
     private final BooleanSupplier anySourceTaskBlocked;
     private final PartitionIdAllocator partitionIdAllocator;
     private final Map<InternalNode, RemoteTask> scheduledTasks;
-    private final Set<Split> pendingSplits = new HashSet<>();
+    private final List<Split> pendingSplits = new ArrayList<>();
```

Comment:
see the line below
```
splitAssignment.values().forEach(pendingSplits::remove); // AbstractSet.removeAll performs terribly here.
```

now that pendingSplits is a list, this will be `List.remove`, which is a slow thing.

Reply:
https://github.com/trinodb/trino/pull/21634